### PR TITLE
[transactions] Implement sorting and filtering on the Transactions and Transaction Details screens

### PR DESCRIPTION
**Pull request by selene-huang**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
- Added sort and filter reducers in `transactionUtils.ts`
- Implemented modals and alerts on the Transactions screen
- Added sort by date and sort by amount functionality on the Transactions screen
- Added filter by date, paid/unpaid, and amount functionality on the Transactions screen
- Added sort by serial number and sort by amount functionality on the Transaction Details screen

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy!"
image
image
image

## How to review
[//]: # "Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?"
Run `npx expo start` and navigate to the Invoices tab. Check that sort and filtering functionality produces the expected results. Click on an invoice and check that sort functionality produces the expected results for the vouchers.

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
The Transaction Details screen design has not yet been fully finalized, so future PRs will need to update its frontend design.

**KNOWN ISSUES:**
- On IOS, you cannot pick the currently selected date on the date pickers in the Filter Modal 

### Online sources
[//]: # "Optional - copy links to any tutorials or documentation that was useful to you when working on this PR"

[This link](https://stackoverflow.com/questions/53048495/does-react-batch-state-update-functions-when-using-hooks) was very useful in helping to debug the displayed arrays and why updates were not displaying immediately.

[This link](https://github.com/react-native-modal/react-native-modal/issues/268) helped to fix a known issue with react-native-modal in which the modal would flicker on closing.

### Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

This PR updates Transactions and Transaction Details screens to v2 versions from https://github.com/calblueprint/vouchers4veggies/pull/54

[//]: # "This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation."
🥦 CC: @sauhardjain

### CODE REVIEW

File: assets/Fonts.tsx [hunk 1]
```diff
@@ -116,6 +123,14 @@ export const DarkGrayText = styled.Text`
   color: ${Colors.darkGray};
 `;
 
+export const MidGrayText = styled.Text`
+  color: ${Colors.midGray};
+`;
+
+export const BlueText = styled.Text`
+  color: #007aff;
```
Comment (allisonhongberkeley) [inline assets/Fonts.tsx]: [nit: line 31] for consistency, define a blue color in Colors?

File: src/screens/transactions/TransactionDetailsScreen.tsx [hunk 2]
```diff
@@ -20,45 +20,79 @@ import {
...
 
-  useEffect(() => {
-    const fetchData = async () => {
-      try {
-        const data = await getTransaction(transactionUuid);
+  const sortButtonText = ['SN', 'SN', 'Date', 'Date'];
+  const sortDescriptionText = [
+    'Serial Number: High to Low',
+    'Serial Number: Low to High',
+    'Date: Newest',
```
Comment (allisonhongberkeley) [inline src/screens/transactions/TransactionDetailsScreen.tsx:60]: I tried sorting by 'Date: Newest' in the app and it doesn't work? Invoices are just randomly sorted.

File: src/components/common/styles.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+export const LeftAlignContainer = styled.View`
+  display: flex;
+  flex: 1;
+  flex-direction: row;
+  justify-content: flex-start;
+  align-content: center;
+`;
+
+export const RightAlignContainer = styled.View`
+  display: flex;
```
Comment (allisonhongberkeley) [inline src/components/common/styles.ts:33]: + flex: 1 for consistently with LeftAlignContainer?

File: src/components/transactions/FilterModal.tsx [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+                    setMaxDatePickerIsVisible(false);
+                    if (e.type === 'set' && e.nativeEvent.timestamp) {
+                      filterDispatch({
+                        type: 'SET_MAX_DATE',
+                        date: new Date(e.nativeEvent.timestamp),
+                      });
+                    }
+                  }}
+                />
+              </DatePickerContainer>
```
Comment (allisonhongberkeley) [inline src/components/transactions/FilterModal.tsx:182]: is there any way to abstract away minDatePicker and maxDatePicker into a separate component? because there's a lot of repetition. same with filter range modals, etc.

File: src/screens/transactions/TransactionDetailsScreen.tsx (hunk 2 above)
Comment (ronniebeggs) [reply src/screens/transactions/TransactionDetailsScreen.tsx:60]: Noticed similar issue

File: src/components/transactions/FilterModal.tsx [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+                <Body1Text>
+                  {filterState.inProgressMinDateIsSet ? (
+                    moment(filterState.inProgressMinDate).format('M/DD/YYYY')
+                  ) : (
+                    <MidGrayText>From</MidGrayText>
+                  )}
+                </Body1Text>
+              </FilterField>
+            </LeftAlignContainer>
+            <HorizontalSpacing />
```
Comment (ronniebeggs) [inline src/components/transactions/FilterModal.tsx]: Can we create a filter-by-date container that positions the date boxes? It'd clean up our styling by removing styled components used only for spacing.

File: src/components/transactions/FilterModal.tsx [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+                    if (e.type === 'set' && e.nativeEvent.timestamp) {
+                      filterDispatch({
+                        type: 'SET_MIN_DATE',
+                        date: new Date(e.nativeEvent.timestamp),
+                      });
+                    }
+                  }}
+                />
+              </DatePickerContainer>
+            )}
```
Comment (ronniebeggs) [inline src/components/transactions/FilterModal.tsx:164]: Currently unable to select the current date in the dropdown.

File: src/components/transactions/FilterModal.tsx [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+                    minAmount: 51,
+                    maxAmount: null,
+                  });
+                }
+              }}
+              minWidth={73}
+            >
+              <Body1Text>50+</Body1Text>
+            </FilterTag>
+          </OneLine>
```
Comment (ronniebeggs) [inline src/components/transactions/FilterModal.tsx:343]: Is it possible to select multiple amount filters? Ex: viewing all invoices between $0-20 by selecting both the `0-10` and `10-20` options. Honestly might be too complex idk.

File: src/components/transactions/SortModal.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+              }}
+            >
+              <ButtonTextWhite>Apply</ButtonTextWhite>
+            </ButtonMagenta>
+          </CenteredContainer>
+        </PaddedScrollView>
+      </SortModalContainer>
+    </Modal>
+  );
+}
```
Comment (ronniebeggs) [inline src/components/transactions/SortModal.tsx:90]: Would it make sense to have a clear field option on the sort modal? Considering that there's likely no default ordering, I'm not sure what it'd revert back to. However, always having a highlighted sort button might give the impression that some invoices are being sorted/shown improperly.

File: src/components/transactions/SortModal.tsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            }}
+          />
+
+          <SortVerticalSpacing />
+          <CenteredContainer>
+            <ButtonMagenta
+              onPress={() => {
+                setIsVisible(false);
+                sortDispatch({ type: 'ON_SUBMIT' });
+              }}
```
Comment (ronniebeggs) [inline src/components/transactions/SortModal.tsx:81]: New invoices aren't rendered until a new filter/sort is applied.

File: src/components/transactions/SortModal.tsx [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+            onPress={() => {
+              setIsVisible(false);
+              sortDispatch({ type: 'RESET_IN_PROGRESS' });
+            }}
+          >
+            <BlueText>
+              <Body1Text>Close</Body1Text>
+            </BlueText>
+          </TouchableOpacity>
+        </CloseButtonContainer>
```
Comment (ronniebeggs) [inline src/components/transactions/SortModal.tsx:62]: When you open the sort modal and apply changes without selecting a sort option, it claims: `Sort by: undefined` within the button. It then appears to sort invoice amounts from high to low.

File: src/components/transactions/styles.ts [hunk 11]
```diff
@@ -19,6 +19,19 @@ export const RightAlignContainer = styled.View`
   margin-right: 29px;
 `;
 
+export const RightAlignContainer = styled.View`
+  flex: 1;
+  width: 100%;
+  flex-direction: row;
+  justify-content: flex-end;
+  align-content: center;
+`;
```
Comment (ronniebeggs) [inline src/components/transactions/styles.ts:20]: Is this redundant styling to have the same component, differing only by its margin?

File: src/components/transactions/styles.ts [hunk 12]
```diff
@@ -85,10 +91,150 @@ export const RedText = styled.Text`
...
+  height: 22px;
+`;
+
+export const FilterVerticalSpacing = styled.View`
+  height: 48px;
+`;
+
+export const HorizontalSpacing = styled.View`
+  width: 20px;
+`;
```
Comment (ronniebeggs) [inline src/components/transactions/styles.ts]: Can we standardize a lot of these styled-components?

**Review by ronniebeggs — CHANGES_REQUESTED** [review]
Looking good. Respect for grinding out this much functionality. Minor bugs and styling nits, but overall coming along nicely.

File: src/components/transactions/SortModal.tsx (hunk 8 above)
Comment (selene-huang) [reply src/components/transactions/SortModal.tsx:90]: The default ordering is date descending (newest first). I don't mind putting in a clear button, though we should probably check the design with Danielle first, but I also don't think it should give the impression of being shown improperly. A sort is just a reordering of the invoices, after all?

File: src/components/transactions/FilterModal.tsx (hunk 7 above)
Comment (selene-huang) [reply src/components/transactions/FilterModal.tsx:343]: I think it's better to keep as is (minimumAmount & maxAmount) because then future devs just have to call the dispatch with these new amounts if they want to add a new filter option. If we enable multi-select, we would have to hardcode every option into the dispatch, and future devs would have to change this on top of the front-end.

File: src/utils/transactionUtils.ts
Comment (sauhardjain) [inline src/utils/transactionUtils.ts:1]: Why are we not using React Context here? I think using Context should reduce some of the redundant code and also increase readability by a lot.

**Review by sauhardjain — APPROVED** [review]
AMAZING WORKKKK @selene-huang! This is such an important PR and major props to you for pushing it out this quickly.

Merging the PR now because we need to have sorting & filtering pushed asap. However, here are some next steps to be addressed in a follow-up PR:

- [ ] Align all elements to the same left margin on `TransactionDetailsScreen`
- [ ] Fix `DatePicker` bug that doesn't allow user to pick the current date
- [ ] Use React Context to improve readability of `trasactionUtils`
- [ ] Add a clear button to the Sort modal
- [ ] Miscellaneous styling and code cleanup
